### PR TITLE
Fix invalid subtype relation in table and anydata

### LINKED ISSUE #32405: No compilation error when calling `toJson()` on a variable of type `table<map<any>>`
**Description:**

Since `toJson()` is only defined on `anydata`, following cases should result in compilation errors.
case i.
```
public function main() {
   table<map<any>> tbl = table [
            {a: "x", b: 3},
            {c: "z", d: 5}
   ];

   json j = tbl.toJson();
}
```
case ii.
```
type Bar record {|
   int id;
   any a;
|};

type Foo record {|
   int id;
   Bar b;
|};

public function main() {
   table<Foo> tb = table[
         {id: 1, b: {id: 1, a: 5}}
   ];

   json j = tb.toJson();

}
```
**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #34696: Incorrect subtype relation with `anydata`
**Description:**
```
// TANY<:A
type A anydata;
type TANY table<map<any>>;
```

`TANY` should not be a subtype of `A`. 
types.isAssignable(TANY, A) should be false.

**Affected Versions:**
swan-lake beta6

### PR DESCRIPTION
**Pull request by KavinduZoysa**
## Purpose

Fixes #34696

Fixes #32405

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal [hunk 1]
```diff
@@ -151,3 +151,14 @@ function testErrorAsAMemberOfAnydata() {
      (anydata|error)[] x = [error("")];
      anydata y = x;
 }
+
+type TANY table<map<any>>;
+type TERR table<map<error>>;
+
+public function testInvalidTableAssignment() {
+    TANY tany = table [ { "a": 2 } ];
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal:159]:
```suggestion
    TANY tany = table [{"a": 2}];
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal [hunk 2]
```diff
@@ -151,3 +151,14 @@ function testErrorAsAMemberOfAnydata() {
...
+    anydata _ = tany; // error
+
+    TERR terror = table [ { "e": error("Message") } ];
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal:162]:
```suggestion
    TERR terror = table [{"e": error("Message")}];
```

**Review by MaryamZi — APPROVED** [review]

**Comment by MaryamZi** [comment]
@KavinduZoysa, BIRSpecTest seems to be failing, which probably means langlib tests are also failing?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal (hunk 1 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal]: Changed

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal (hunk 2 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal]: Changed

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java [hunk 3]
```diff
@@ -68,6 +68,9 @@ private boolean isAnydata(BType type) {
             case TypeTags.JSON:
             case TypeTags.XML:
             case TypeTags.XML_TEXT:
+            case TypeTags.XML_ELEMENT:
+            case TypeTags.XML_COMMENT:
+            case TypeTags.XML_PI:
```
Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java:73]: Why was this change needed? Shouldn't the same be done in `IsAnydataUniqueVisitor`? 

`IsPureTypeUniqueVisitor` and pure types (what used to be `anydata|error`) are no longer applicable and we should be able to get rid of this. Shall we at least get rid of the usage of this in `IsAnydataUniqueVisitor` for now?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java:73]: Is there a test covering whatever this change was required for?

Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java:73]: This change was done because of a test failure. I did not know that we are trying to get rid of `IsPureTypeUniqueVisitor`. I also noticed that there are not many usages of `IsPureTypeUniqueVisitor`. Shall I send a separate PR to remove `IsPureTypeUniqueVisitor`?

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java:73]: Shall we remove it from `IsAnydataUniqueVisitor` in this PR itself, because fixing the original issue required changes in `IsPureTypeUniqueVisitor`? Just replacing them with recursive calls to `IsAnydataUniqueVisitor` itself should work, right?

Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java:73]: Changed

Comment (MaryamZi) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java:73]: Do we still need this change?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal [hunk 4]
```diff
@@ -151,3 +151,14 @@ function testErrorAsAMemberOfAnydata() {
      (anydata|error)[] x = [error("")];
      anydata y = x;
 }
+
+type TANY table<map<any>>;
```
Comment (MaryamZi) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal:155]: Do we already have runtime tests?

**Review by MaryamZi — COMMENTED** [review]
@KavinduZoysa, we need to rebase this PR to https://github.com/ballerina-platform/ballerina-lang/tree/ballerina-2.0.x.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java (hunk 3 above)
Comment (KavinduZoysa) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsPureTypeUniqueVisitor.java]: Removed

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal (hunk 4 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_negative_test.bal:155]: Add tests

**Comment by MaryamZi** [comment]
Changes LGTM, but we need to decide on which branch we should merge this to.

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_test.bal [hunk 5]
```diff
@@ -554,10 +554,10 @@ function testAnydataToMap() {
         convertedMCfoo = ad;
     }
 
-    map<table<map<any>>> mt = {};
+    map<table<map<anydata>>> mt = {};
     ad = mt;
     map<table<map<any>>> convertedMt;
-    if (ad is map<table<map<any>>>) {
+    if (ad is map<table<map<anydata>>>) {
```
Comment (gabilang) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_test.bal:560]: Shall we add some more negative test cases for type-test expression checks? 
eg.
```ballerina
table<map<any>> t = table [];
map<table<map<any>>> mt = {};
assertFalse(t is anydata);
assertFalse(mt is anydata); 
```

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_test.bal [hunk 6]
```diff
@@ -880,6 +882,8 @@ function testRuntimeIsAnydata() {
     assertFalse(g is anydata);
     assertFalse(h is anydata);
     assertFalse(i is anydata);
+    assertFalse(j is anydata);
+    assertFalse(k is anydata);
```
Comment (gabilang) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_test.bal:886]:
```
assertFalse(table [{"e": error("Message")}] is anydata);
assertFalse(table [{"a": {"x": error("Message")}}] is anydata);
```
Above is-expr checks provide compilation errors with the fix. But ideally it shouldn't right?

**Comment by MaryamZi** [comment]
@KavinduZoysa, can you mention https://github.com/ballerina-platform/ballerina-lang/issues/32405 also as a fixed issue?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_test.bal (hunk 5 above)
Comment (KavinduZoysa) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/anydata/anydata_test.bal:560]: Still, we are getting a compiler error for this case. 
It should be fixed via https://github.com/ballerina-platform/ballerina-lang/issues/24451

**Comment by KavinduZoysa** [comment]
> @KavinduZoysa, can you mention #32405 also as a fixed issue?

Done

**Review by gimantha — APPROVED** [review]
LGTM

**Comment by chiranSachintha** [comment]
The following scenario should also result in an error. right?
```ballerina
public function main() {
        table<map<any>> tab3 = table [
          {id: 12, name: "abc"},
          {id: 34, name: "def"}
    ];
    
    anydata l = <anydata> tab3;
}
```

**Comment by MaryamZi** [comment]
> The following scenario should also result in an error. right?
> 
> ```ballerina
> public function main() {
>         table<map<any>> tab3 = table [
>           {id: 12, name: "abc"},
>           {id: 34, name: "def"}
>     ];
>     
>     anydata l = <anydata> tab3;
> }
> ```

This should panic.

I remember checking the equivalent of `<any> tab3 is anydata`, when testing the PR initially, to verify the runtime behaviour. This returned/returns false as expected.

Also, `<anydata> <any> tab3` panics.

So I suspect this may be getting optimized as maybe a NOOP at compile-time. @KavinduZoysa, let's check and fix this in this PR.

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java [hunk 7]
```diff
@@ -70,7 +70,10 @@ private boolean isAnydata(BType type) {
             case TypeTags.BOOLEAN:
             case TypeTags.JSON:
             case TypeTags.XML:
-            case TypeTags.TABLE:
+            case TypeTags.XML_TEXT:
+            case TypeTags.XML_ELEMENT:
```
Comment (chiranSachintha) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:74]: Is it correct to add `JSON` here? (In L71)(not related to this issue)
By that, we allow the following kind of scenario also.
```ballerina
public function main() {
        table<map<json>> tab3 = table [ ];
    
        anydata l = tab3;
}
```

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:74]: Why should this be disallowed?

Comment (chiranSachintha) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:74]: is `table<map<json>>` is sub type of `anydata`. ?
Spec says `anydata can be defined as () | boolean | int | float | decimal | string | XML | anydata[] | map<anydata> | table<map<anydata>>` (Is `json` implies from these types)

Comment (MaryamZi) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/IsAnydataUniqueVisitor.java:74]: `json` is `() | boolean | int | float | decimal | string | json[] | map<json>`.

`anydata` is `() | boolean | int | float | decimal | string | xml | anydata[] | map<anydata> | table<map<anydata>>`.

The set of shapes represented by `json` is a subtype of that represented by `anydata`. Therefore `json` is a subtype of `anydata`.

**Comment by KavinduZoysa** [comment]
> 

@MaryamZi, are we expecting to panic below example?

```ballerina
public function main() {
    map<any> m = { "a": 2, "b": 3 };
    anydata _ = <anydata> m; // ?
}
```

**Comment by MaryamZi** [comment]
> > 
> 
> @MaryamZi, are we expecting to panic below example?
> 
> ```ballerina
> public function main() {
>     map<any> m = { "a": 2, "b": 3 };
>     anydata _ = <anydata> m; // ?
> }
> ```

Your thoughts please?

**Comment by KavinduZoysa** [comment]
> > > 
> > 
> > 
> > @MaryamZi, are we expecting to panic below example?
> > ```ballerina
> > public function main() {
> >     map<any> m = { "a": 2, "b": 3 };
> >     anydata _ = <anydata> m; // ?
> > }
> > ```
> 
> Your thoughts please?

I think this should panic, because m is mutable. Even if the current value belongs to anydata it can be changed.

In the current implementation, this is working without panicing.
So I think the issue discussed in https://github.com/ballerina-platform/ballerina-lang/pull/34803#issuecomment-1070375485 occurs in `map` too.

**Comment by MaryamZi** [comment]
> > > > 
> > > 
> > > 
> > > @MaryamZi, are we expecting to panic below example?
> > > ```ballerina
> > > public function main() {
> > >     map<any> m = { "a": 2, "b": 3 };
> > >     anydata _ = <anydata> m; // ?
> > > }
> > > ```
> > 
> > 
> > Your thoughts please?
> 
> I think this should panic, because m is mutable: even if the current value belongs to anydata it can be changed. Also this `m is anydata` returns `false`.
> 
> In the current implementation, this is working without panicing. So I think the issue discussed in [#34803 (comment)](https://github.com/ballerina-platform/ballerina-lang/pull/34803#issuecomment-1070375485) occurs in `map` too.
> 
> @MaryamZi, please correct me if I am wrong.

Yes, this should panic.

**Review by hasithaa — APPROVED** [review]
